rust/origen/src/services/arm_debug/jtag_dp.rs

```diff
@@ -0,0 +1,300 @@
+            bc.set_data(BigUint::from(0 as u8));
+
+            let bc = get_bc_for!(dut, reg, "a")?;
+            bc.set_data(BigUint::from((transaction.addr()? & 0xF) >> 2 as u8));
+
+            let bc = get_bc_for!(dut, reg, "rnw")?;
+            bc.set_data(BigUint::from(1 as u8));
+
+            let bc = reg.bits(dut);
+
+            self.write_register(dut, services, bc.to_write_transaction(&dut)?)?;
+        } else {
+            self.update_ir(&dut, &services, self.dpacc_select as u32)?;
+        }
+
+        let mut t = transaction.clone();
+        t.prepend_data(BigUint::from(0 as u8), 3)?;
+        t.bit_enable = t.bit_enable & BigUint::from(0x7_FFFF_FFF8 as u64);
+        let jtag_service = services.get_service(self.jtag_id)?;
+        let jtag = jtag_service.as_jtag()?;
+        jtag.verify_dr(&dut, &t)?;
+        TEST.close(n_id)?;
+        Ok(())
+    }
+
+    pub fn write_ap(
+        &self,
+        dut: &MutexGuard<Dut>,
+        services: &Services,
+        transaction: Transaction,
+    ) -> Result<()> {
+        let n_id = TEST.push_and_open(node!(JTAGDPWriteAP, self.id, transaction.clone(), None));
+        let reg = get_reg!(dut, self.address_block_id, "apacc");
+
+        let bc = get_bc_for!(dut, reg, "a")?;
+        bc.set_data(BigUint::from((transaction.addr()? & 0xF) >> 2 as u8));
+
+        let bc = get_bc_for!(dut, reg, "data")?;
+        bc.set_data(transaction.data);
+
+        let bc = get_bc_for!(dut, reg, "rnw")?;
+        bc.set_data(BigUint::from(0 as u8));
+
+        let bc = reg.bits(dut);
+        self.write_register(dut, services, bc.to_write_transaction(&dut)?)?;
+
+        TEST.close(n_id)?;
+        Ok(())
+    }
+
+    pub fn verify_ap(
+        &self,
+        dut: &MutexGuard<Dut>,
+        services: &Services,
+        transaction: Transaction,
+        write_apacc: bool,
+    ) -> Result<()> {
+        let n_id = TEST.push_and_open(node!(JTAGDPVerifyAP, self.id, transaction.clone(), None));
+        if write_apacc {
+            let reg = get_reg!(dut, self.address_block_id, "apacc");
+
+            let bc = get_bc_for!(dut, reg, "data")?;
+            bc.set_data(BigUint::from(0 as u8));
+
+            let bc = get_bc_for!(dut, reg, "a")?;
+            bc.set_data(BigUint::from((transaction.addr()? & 0xF) >> 2 as u8));
+
+            let bc = get_bc_for!(dut, reg, "rnw")?;
+            bc.set_data(BigUint::from(1 as u8));
+
+            let bc = reg.bits(dut);
+
+            self.write_register(dut, services, bc.to_write_transaction(&dut)?)?;
+        } else {
+            self.update_ir(&dut, &services, self.apacc_select as u32)?;
+        }
+
+        let mut t = transaction.clone();
+        t.prepend_data(BigUint::from(0 as u8), 3)?;
+        t.bit_enable = t.bit_enable & BigUint::from(0x7_FFFF_FFF8 as u64);
+        let jtag_service = services.get_service(self.jtag_id)?;
+        let jtag = jtag_service.as_jtag()?;
+        jtag.verify_dr(&dut, &t)?;
```

Comment:
The ARM debug interface AS specifies a 35 bit write of 0x7 to perform the read operation of DRW and the subsequent shift out of the register data. We've been getting away with shifting in all 0's, hasn't caused any issues.

Reply:
I referenced both, but when I ran into errors, I deferred back to the O1 jtag driver, as I could generate waves with that to actually look for differences in the state.

For that particular line though, that's just setting the enables, not the data. Its "don't-caring" the previous acknowledgment.